src/util/address.rs

```diff
@@ -145,6 +154,173 @@ impl FromStr for AddressType {
+/// transaction output for transactions starting with opcodes ranging from 0
+/// to 16 (inclusive).
+///
+/// Structure helps to limit possible version of the witness according to the
+/// specification; if a plain `u8` type will be used instead it will mean that
+/// version > 16, which is incorrect.
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+#[repr(u8)]
+pub enum WitnessVersion {
+    /// Initial version of Witness Program. Used for P2WPKH and P2WPK outputs
+    V0 = 0,
+    /// Version of Witness Program used for Taproot P2TR outputs
+    V1 = 1,
+    /// Future (unsupported) version of Witness Program
+    V2 = 2,
+    /// Future (unsupported) version of Witness Program
+    V3 = 3,
+    /// Future (unsupported) version of Witness Program
+    V4 = 4,
+    /// Future (unsupported) version of Witness Program
+    V5 = 5,
+    /// Future (unsupported) version of Witness Program
+    V6 = 6,
+    /// Future (unsupported) version of Witness Program
+    V7 = 7,
+    /// Future (unsupported) version of Witness Program
+    V8 = 8,
+    /// Future (unsupported) version of Witness Program
+    V9 = 9,
+    /// Future (unsupported) version of Witness Program
+    V10 = 10,
+    /// Future (unsupported) version of Witness Program
+    V11 = 11,
+    /// Future (unsupported) version of Witness Program
+    V12 = 12,
+    /// Future (unsupported) version of Witness Program
+    V13 = 13,
+    /// Future (unsupported) version of Witness Program
+    V14 = 14,
+    /// Future (unsupported) version of Witness Program
+    V15 = 15,
+    /// Future (unsupported) version of Witness Program
+    V16 = 16,
+}
+
+impl fmt::Display for WitnessVersion {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", *self as u8)
+    }
+}
+
+
+impl FromStr for WitnessVersion {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let version = u8::from_str(s).map_err(|err| Error::UnparsableWitnessVersion(err))?;
+        WitnessVersion::from_no(version)
+    }
+}
+
+impl WitnessVersion {
+    /// Converts 5-bit unsigned integer value matching single symbol from Bech32(m) address encoding
+    /// ([`bech32::u5`]) into [`WitnessVersion`] variant.
+    ///
+    /// # Returns
+    /// Version of the Witness program
+    ///
+    /// # Errors
+    /// If the integer does not corresponds to any witness version, errors with
+    /// [`Error::InvalidWitnessVersion`]
+    pub fn from_u5(value: ::bech32::u5) -> Result<Self, Error> {
+        WitnessVersion::from_no(value.to_u8())
+    }
+
+    /// Converts 8-bit unsigned integer value into [`WitnessVersion`] variant.
+    ///
+    /// # Returns
+    /// Version of the Witness program
+    ///
+    /// # Errors
+    /// If the integer does not corresponds to any witness version, errors with
+    /// [`Error::InvalidWitnessVersion`]
+    pub fn from_no(no: u8) -> Result<Self, Error> {
```

Comment:
Nit: use `num` instead of `no` for a number variable? Also, this function would probably be better named `from_u8`

Reply:
It was originally named `from_u8`, but later I decided its a bad name, since there are two different u8-based representations for witness version: by script byte (`0`, `0x51`-`0x60`) and as version number (incremental from 0 to 16). So while I'm fine with using `from_num` instead of `from_no`, it will be confusing to use `from_u8`.

Reply:
Possibly `raw_num` or `raw_u8`?

Reply:
+1 for num instead of no